indexer-api/src/infra/api/v1/subscription/unshielded.rs

```diff
@@ -80,114 +83,177 @@ where
-                event
-            });
+        let storage = cx.get_storage::<S>();
+        let subscriber = cx.get_subscriber::<B>();
 
-        let encoded_address_for_progress = encoded_address.clone();
-        let progress_updates = progress_updates::<S>(cx, address)
-            .await?
-            .map_ok(move |event| {
-                debug!(address = encoded_address_for_progress; "emitting PROGRESS event");
-                event
-            });
+        // Use 0 as default to include all transactions from genesis.
+        // Since transaction IDs start from 1 (BIGSERIAL/AUTOINCREMENT), using >= 0
+        // ensures we include the genesis transaction (ID 1) and all subsequent ones.
+        let from_transaction_id = transaction_id.unwrap_or(0);
+        let utxo_indexed_events = subscriber.subscribe::<UnshieldedUtxoIndexed>();
 
-        let events = tokio_stream::StreamExt::merge(update_events, progress_updates);
+        let updates = {
+            let address = address.clone();
 
-        Ok(events)
-    }
-}
+            try_stream! {
+                debug!(
+                    address:?,
+                    from_transaction_id;
+                    "starting unshielded subscription with historical replay"
+                );
 
-#[trace(properties = { "address": "{address:?}" })]
-async fn unshielded_updates<'a, S, B>(
-    cx: &'a Context<'a>,
-    address: ByteVec,
-    network_id: NetworkId,
-) -> async_graphql::Result<
-    impl Stream<Item = async_graphql::Result<UnshieldedUtxoEvent<S>>> + use<'a, S, B>,
->
-where
-    S: Storage,
-    B: Subscriber,
-{
-    let storage = cx.get_storage::<S>();
-    let subscriber = cx.get_subscriber::<B>();
-
-    let address_for_filter = address.clone();
-    let utxo_indexed_events = subscriber
-        .subscribe::<UnshieldedUtxoIndexed>()
-        .try_filter(move |event| ready(event.address == address_for_filter));
-
-    let updates = try_stream! {
-        let mut utxo_indexed_events = pin!(utxo_indexed_events);
-        while let Some(UnshieldedUtxoIndexed { address: _, transaction_id }) = utxo_indexed_events
-            .try_next()
-            .await
-            .internal("get next UnshieldedUtxoIndexed event")?
-        {
-            debug!(
-                address:?,
-                transaction_id;
-                "handling UnshieldedUtxoIndexed event"
-            );
-
-            let tx = storage
-                .get_transaction_by_id(transaction_id)
-                .await
-                .internal("fetch tx for subscription event")?;
-
-            let tx = match tx {
-                Some(tx) => tx,
-                None => {
-                    warn!(transaction_id; "transaction not found, skipping event");
-                    continue;
+                // Phase 1: Replay all historical transactions for this address.
+                let historical_transactions = storage
+                    .get_transactions_involving_unshielded(&address)
```

Comment:
This method needs a `from_id` parameter. Then you can skip the below "skip" section which is very inefficient.